[dynamo][guard refactor] C++ Guard data structure

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110590
* __->__ #108839

**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

anijain2305 committed:
wip

[ghstack-poisoned]

anijain2305 committed:
Update on "wip"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "wip"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "wip"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "wip"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "wip"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "wip"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "wip"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "wip"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "wip"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
+    this->failure_reason = failure_reason;
+  }
+
+  std::string repr() {
```
this is not very C++, can we reload the stream operator?

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
+ * failure. The data structure is also accessible in Python.
+ */
+struct GuardDebugInfo {
+  GuardDebugInfo(int num_guards_executed, std::string failure_reason) {
```
nit: std::string_view

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
+class LeafGuard {
+ public:
+  // Runs the guard and prepares a GuardDebugInfo object.
+  std::pair<bool, GuardDebugInfo> check_with_debug_info(py::object value) {
```
Why are you taking the arg by value?

Also, in C++, extremely unusual to do a bool pair. Instead, I would make the object implement the bool operator to represent pass/fail, or do it via static typing.

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
+    bool result = run_guards(value);
+    if (result == false) {
+      std::string reason = get_failure_reason(value);
+      return std::make_pair(result, GuardDebugInfo(0, reason));
```
You are not setting num_guards_executed?

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
+  std::pair<bool, GuardDebugInfo> check_with_debug_info(py::object value) {
+    bool result = run_guards(value);
+    if (result == false) {
+      std::string reason = get_failure_reason(value);
```
const ref auto, unless you are guessing at RVO, in which case, don't ;)

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
+  }
+
+  // Runs the guard
+  bool check(py::object value) {
```
I get that `bool check` here and on GuardManager are meant to match, but i'd rather we just get rid of this extra fn and call run_guards. This isn't doing anything but calling that!

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
+ * guards of type PythonLambaGuard and incrementally move expensive guards to
+ * C++.
+ */
+class PythonLambdaGuard : public LeafGuard {
```
Runtime polymorphism in C++ is unusual, imo, compared to compile time.

If all a PythonLambdaGuard is doing is defining 2 functions, consider a template. Its kind of a personal bias, but I *strongly* prefer templated code. 

```
template <typename CheckFunc, typename FailureFunc>
class LeafGuard {
public:
    LeafGuard(CheckFunc check_fn, FailureFunc failure_fn)
        : _check_fn(check_fn), _failure_fn(failure_fn) {}

    std::pair<bool, GuardDebugInfo> check_with_debug_info(py::object value) {
        bool result = run_guards(value);
        if (!result) {
            std::string reason = get_failure_reason(value);
            return std::make_pair(result, GuardDebugInfo(0, reason));
        }
        return std::make_pair(result, GuardDebugInfo(0, "PASS"));
    }

    bool check(py::object value) {
        return run_guards(value);
    }

    bool run_guards(py::object value) {
        return _check_fn(value);
    }

    std::string get_failure_reason(py::object value) {
        return _failure_fn(value);
    }

private:
    CheckFunc _check_fn;
    FailureFunc _failure_fn;
};

auto check_lambda = [](py::object value) -> bool { /* ... */ };
auto failure_lambda = [](py::object value) -> std::string { /* ... */ };

LeafGuard<decltype(check_lambda), decltype(failure_lambda)> guard(check_lambda, failure_lambda);
```

voznesenskym replied:
SGTM, author's choice :)

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
+    _attr_name = name;
+  }
+
+  bool equals(py::str name) const {
```
c++ equality operator overload

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
+ * Base class representing how to access the child value from the py::object
+ * given to the parent check function.
+ *
+ * These will have one to one mapping with Source objects in Dynamo. Source
+ * provides a way to access the underlying object using its name() function.
+ * GuardAccessor is basically its C++ tree equivalent.
```
Why do we need C++ dynamo sources? I was thinking we don't need sources in C++... since they are just used for guard construction. Once the guard FN is made, you can just pass it down to C++ as a string.

voznesenskym commented:
I started reviewing the code and hit a bunch of nits, but then stopped because I found a directional stumbling block. 

This PR seems to want to move a lot of guard creation components (sources, for example) over to C++. 

I was thinking we don't need to do that, at least, for guard trees or the priority queue. However, if we want to run guards entirely in C++, I think this will take more design.

Before we get into the nitty gritty, and I apologize for asking this, because I always feel like such a bore doing it - but is there a doc anywhere on what the arch here is? I'm having a little trouble zooming out and understanding (1) The goal (is it guard trees? Getting all guards to C++ entirely? (2) how the pieces here fit together and (3) how it will fit together with the existing guard code. 

Sorry for distracting here and asking for a doc or detailed overview, but I'd like us to iron out a design first and then jump into the impl.

voznesenskym reviewed (changes requested):
I started reviewing the code and hit a bunch of nits, but then stopped because I found a directional stumbling block.

This PR seems to want to move a lot of guard creation components (sources, for example) over to C++.

I was thinking we don't need to do that, at least, for guard trees or the priority queue. However, if we want to run guards entirely in C++, I think this will take more design.

Before we get into the nitty gritty, and I apologize for asking this, because I always feel like such a bore doing it - but is there a doc anywhere on what the arch here is? I'm having a little trouble zooming out and understanding (1) The goal (is it guard trees? Getting all guards to C++ entirely? (2) how the pieces here fit together and (3) how it will fit together with the existing guard code.

Sorry for distracting here and asking for a doc or detailed overview, but I'd like us to iron out a design first and then jump into the impl.

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,371 @@ static struct PyModuleDef _module = {
+};
+
+/**
+ * GuardManger encapsulates all the guards related to a particular py::object.
```
Manager

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -622,5 +987,82 @@ PyObject* torch_c_dynamo_guards_init() {
+  // Guard Manager
+  py::class_<GuardManager, std::unique_ptr<GuardManager>>(py_m, "GuardManager")
+      .def(py::init<>())
+      .def("check", &GuardManager::check)
```
Another optimization I would like to see here is calling guards directly via C++ rather than round-tripping through Python.  If we do this, then this check method doesn't need to be bound here. 

The goal here is the pybind should only be used for guard construction, not guard checking.

anijain2305 replied:
There is fast path now - `check_pybind`. We will use this when we stick the GuardManager in cache entry. For debugging purpose, I still have a `check` function exposed to Python.

ezyang commented:
you're well provisioned for reviewers, so imma bow out. i'll read through the implementation once it lands

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,371 @@ static struct PyModuleDef _module = {
+ */
+class PythonLambdaGuard : public LeafGuard {
+ public:
+  PythonLambdaGuard(py::object guard_check_fn, py::object print_failure_fn) {
```
@animesh if you want to read more about how and why https://en.cppreference.com/w/cpp/language/constructor

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,371 @@ static struct PyModuleDef _module = {
+
+ private:
+  // The user provided lambda function for check_fn.
+  py::function _guard_check_fn;
+  // The user provided lambda function to get guard failure reason.
+  py::function _print_failure_fn;
```
I don't understand the ownership semantics of holding something like this

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,371 @@ static struct PyModuleDef _module = {
+    return py::cast<bool>(_guard_check_fn(value));
+  }
+
+  std::string get_failure_reason(py::object value) override {
+    return py::cast<std::string>(_print_failure_fn(value));
+  }
```
string_view, here and everywhere

anijain2305 replied:
So I tried `string_view` and got some issues accessing in Python using pybind - https://zpz.github.io/blog/python-cpp-pybind11-stringview/

Though I can work harder to fix the `string_view` + pybind issue, but I don't think we need to complicate it. The `std::string` part of the codebase is not in the hot path. We need string only for recompilation reasoning which we will soon move to convert_frame.py - https://github.com/pytorch/pytorch/issues/108950

So I am going to not implement string_view here.

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,371 @@ static struct PyModuleDef _module = {
+    // failing on the first accessor itself. This is helpful when we have
+    // already sorted the guards once, and dont need to sort again.
+    if (result == false and failed_on_first == false) {
+      // Inplace sort the child guards by fail count. This moves the guard with
+      // higher fail count earlier in the queue, and enables fail fast for the
+      // next check_with_debug_info.
+
+      // An alternate implementation was to use priority queue directly on
+      // _accessors, but it was rejected because of the complexity of
+      // popping and creating a new pq on each run_guards. Moreover, this sort
+      // is happening on the unhappy path when check_with_debug_info guard
+      // fails. So, its probably ok.
+      std::sort(
+          _accessors.begin(),
+          _accessors.end(),
+          [](const std::unique_ptr<GuardAccessor>& a,
+             const std::unique_ptr<GuardAccessor>& b) {
+            return a->get_guard_manager()->fail_count() >=
+                b->get_guard_manager()->fail_count();
+          });
```
Discussed offline, we're all cool with it

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,371 @@ static struct PyModuleDef _module = {
+    return _fail_count;
+  }
+
+  // Returning raw pointers because we can't return unique_ptr and pybind does
```
I don't get it, this is ownership ambiguity. Returning unique_ptr would mean you are passing ownership. If you have firm guarantees about lifecycle, return const refs?

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,465 @@ static struct PyModuleDef _module = {
+
+ private:
+  // The user provided lambda function for check_fn.
+  py::function _guard_check_fn;
```
Lost in squashing commits.

@voznesenskym  says "I don't understand the ownership semantics of holding something like this"

anijain2305 replied:
IIUC, I am using py::object in the constructor, and then copying it in the this py::function, which increases a ref count. I don't think ownership changes. Its just that `LambdaGuard` object holds a reference to the python callables - https://pybind11.readthedocs.io/en/stable/reference.html#with-reference-counting


@jansel correct me if I am wrong.

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,465 @@ static struct PyModuleDef _module = {
+    bool result = check_nopybind(value);
+    std::string failure_reason = "";
+    if (!result) {
+      failure_reason = get_failure_reason(value);
```
Lost in squashing commits

@voznesenskym says "const ref auto, unless you are guessing at RVO, in which case, don't ;)"

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,371 @@ static struct PyModuleDef _module = {
+  std::string failure_reason;
+
+  // Total number of executed guards so far.
+  int num_guards_executed;
```
This is for debugging guard shuffling. It refers to the number of guards executed so far in the guard tree.

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,459 @@ static struct PyModuleDef _module = {
+      : GuardAccessor(name), _attr_name(name.ptr()) {}
+
+  PyObject* access(PyObject* obj) const override { // borrowed ref
+    return PyObject_GetAttr(obj, _attr_name);
```
`PyObject_GetAttr` returns a new reference. So, should we probably DECREF it somwhere?

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,459 @@ static struct PyModuleDef _module = {
+      : GuardAccessor(name), _attr_name(name.ptr()) {}
+
+  PyObject* access(PyObject* obj) const override { // borrowed ref
+    return PyObject_GetItem(obj, _attr_name);
```
`PyObject_GetItem` returns a new reference. So, should we probably DECREF it somewhere?

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,459 @@ static struct PyModuleDef _module = {
+      : GuardAccessor(name), _attr_name(name.ptr()) {}
+
+  PyObject* access(PyObject* obj) const override { // borrowed ref
+    return PyDict_GetItem(obj, _attr_name);
```
`PyDict_GetItem` returns a borrowed ref, so this is probably ok. Please double check.

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+  }
+
+  virtual ~GuardAccessor() {
+    // _accessed_value is owned by the GuardAccessor. method access returns an
+    // owning reference.
```
Fix the comment. method access returns a borrowed ref.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+  GuardDebugInfo(
+      bool result,
+      std::string failure_reason,
+      int num_guards_executed) {
```
use `: result(...), ... {}` style constructors.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+
+  // Runs the lambda function with the current f_locals value.
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    return PyObject_IsTrue(PyObject_CallOneArg(_guard_check_fn_pyobj, value));
```
I think we are leaking a reference to Py_True/Py_False here. PyObject_CallOneArg returns an owned object.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+ */
+class NoTensorAliasingGuard : public LeafGuard {
+ public:
+  NoTensorAliasingGuard() : _is_first_call(true) {}
```
How does this get reset back to false after the guards run?

We might need a separate GuardCheckState object (that only lives while checks are running).

anijain2305 replied:
I added a virtual method - `reset_state_on_guard_failure` on the relational guards. This is the flow

1) The GuardManager has one more private member now - `_relational_guard_resetters`, a vector or Relational Guard pointers.
2) While installing a no_tensor_aliasing guard - we find the lowest common ancestor (lca) of the two guard managers. The lca node has the responsibility to call the `reset_state_on_guard_failure` of the newly added guard. This is done by registering the `_relational_guard_resetters` on the lca node.
3) On guard failure, GuardManager iterates through the resetters and call the above method.

You can review beginning from `install_no_tensor_aliasing_guard` method.

anijain2305 replied:
Discussed offline. Resolving this for now.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+  }
+
+  std::string get_failure_reason(PyObject* value) override { // borrowed ref
+    return "";
```
?

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+ public:
+  GuardAccessor(py::object accessor_key)
+      : _guard_manager(std::make_unique<GuardManager>()),
+        _accessor_key(accessor_key) {}
```
```suggestion
        _accessor_key(std::move(accessor_key)) {}
```
to avoid an extra incref/decref.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+  // accessor key could be py::str for getattr, getitem or py::function for
+  // lambda accessor.
+  py::object _accessor_key;
+  // Keeps the pointer to the accessed value to implement RAII mechanism.
+  PyObject* _accessed_value;
```
How does this work accross multiple guard checks?

I'd expect this object to live forever, not be created/destroyed on each check.

anijain2305 replied:
This is the accessed_value for the given f_locals. So, this should not live forever. This should change between each check. Does that make sense?

jansel replied:
But this is not defined to something per-f_locals.  It is defined on part of the guard data stucture.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+
+  PyObject* access(PyObject* obj) override { // borrowed ref
+    PyObject* ret = PyObject_GetAttr(obj, _attr_name);
+    // Store the owning reference to _accessed_value to be decref'd on
+    // destruction.
+    _accessed_value = ret;
```
Ownership model seems wrong here.
1) This object won't get destroyed after guards run.
2) If _accessed_value already has a value in it, that value will get leaked.

anijain2305 replied:
The value is decref'd in the destructor of GuardAccessor. 
~~~
virtual ~GuardAccessor() {
    // _accessed_value is owned by the GuardAccessor. method access returns an
    // owning reference.
    Py_DECREF(_accessed_value);
  } 
~~~

The idea is that when we walk through the accessors in the tree, the GuardAccssor object going out of scope will decrement the ref count of the _accessed_value (in the following loop).

~~~
    bool failed_on_first = true;
    for (const auto& accessor : _accessors) {
      auto& manager = accessor->get_guard_manager();
      result = result && manager->check_nopybind(accessor->access(value));
      if (!result) {
        break;
      }
      failed_on_first = false;
    }
~~~

I am happy to discuss offline if there is a better way to represent this.

jansel replied:
`const auto& accessor` means you take a *reference* to the accessor, so it will not be destroyed when it goes out of scope.

You also don't want to be creating/destroying the guard accessor each time, because it holds ownership of the keys (that you want to persist).

I think it might be cleaner to make the sub-guards a property of the accessor, and have access just use a `check()` method that returns a bool.  That way ownership doesn't need to cross function boundaries.

anijain2305 replied:
Thanks for catching this. It was silly on my end. 

I liked your idea about running subguards from accessor itself. I pursued this and found that calling guard_manager->check from inside of GuardAccessor will create a circular dependency that cannot be met by forward declaration in the cpp file (we already forward declare GuardManger in the cpp file today). I could move things to header file etc, but not sure if this is what we want either.

Other alternative is to keep the current structure same and `access` always return a new reference. So, the Guardmanager code will look something like this

~~~
    bool failed_on_first = true;
    for (const auto& accessor : _accessors) {
      auto& manager = accessor->get_guard_manager();
      PyObject* x  = accessor->access(value); // new ref
      result = result && manager->check_nopybind(x);
      Py_DECREF(x);
      if (!result) {
        break;
      }
      failed_on_first = false;
    }

~~~

The pro here is all code is in one file so simpler to read. Con is that we will have to unnecessary increment a ref count for PyDict_GetItem accessor because that one does not return a borrowed ref. We could make helper functions - `maybe_decref` to avoid that as well. But then its a vtable function call, and just inc/dec the ref count should be faster (not slower) than the vtable call.

Let me know if what are your opinions on this.

jansel replied:
Circular deps should not be an issue.  Just forward declare in the same file.  Headers versus cpp files doesn't matter in C++, they are all just one file after the `#include`'s are processed.  So anything that can go in a header can go in a cpp.

anijain2305 replied:
Ok, it works. I just had to move the derived classes of GuardAccessor to the end.

Thanks for the suggestion. The ownership does not cross function boundaries now. Its cleaner now.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+
+  // GuardManager is the owner of the leaf_guards
+  void add_leaf_guard(std::shared_ptr<LeafGuard> leaf_guard) {
+    _leaf_guards.push_back(leaf_guard);
```
```suggestion
    _leaf_guards.emplace_back(std::move(leaf_guard));
```

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+   * already present, we just return the guard manager.
+   */
+  template <typename GuardAccessorT>
+  GuardManager* get_child_manager(py::object accessor_key) {
```
```suggestion
  GuardManager* get_child_manager(const py::object& accessor_key) {
```
incref seems unnessary.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+    bool result = true;
+    // Iterate over leaf guards
+    for (const auto& guard : _leaf_guards) {
+      result = result && guard->check_nopybind(value);
```
we should early exit if a guard fails, no need for the sorting stuff or fail_count

anijain2305 replied:
1) I think we need to increment the fail_count. Yeah, sorting of the children is not needed.
2) result && will short-circuit and will not run if result is False.

jansel replied:
You should exit the loop early and just return.  This is what check_verbose does.

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented:
@jansel this is ready for next round of review.

jansel reviewed (commented):
High level the changes seem good.  I'd like to see actual usage of this (perhaps in another PR in the stack) since I am sure that will expose more issues.

anijain2305 commented:
@jansel Do you mean that we should land this PR only after we have some confidence that it works end to end? I am expecting a decent number of small PRs on top of this to actually make it work end to end. I don't think waiting for all those PRs is the right way. 

Other way is that I can create a stacked wip pr that shows promise on few examples, validating this PR so that we can land it. Let me know your thoughts.

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guard refactor] C++ Guard data structure"


**This PR does not change any existing guards flow in Dynamo.**

This PR is introducing new C++ level data structure to holds guard. Goal is to have necessary python extensions to conceptually represent any source and any guard using these data structures. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
